File: zosciimq/nodejs/replikate.js
// Cyborg ZOSCII MQ v20251030
// (c) 2025 Cyborg Unicorn Pty Ltd.
// This software is released under MIT License.

// ZOSCII MQ Server-to-Q Replikate (replikate.js)
// 
// Pulls messages from a URL and publishes them to a local Target Queue.
// State is managed using a unique file per target queue, making it concurrency-safe.
// 
// Execution Example:
// node replikate.js --url=http://other.server/index.php --sq=sourcequeue --tq=targetqueue

const fs = require('fs');
const path = require('path');
const http = require('http');
const https = require('https');
const url = require('url');
const querystring = require('querystring');
const { promisify } = require('util');

const readFile = promisify(fs.readFile);
const writeFile = promisify(fs.writeFile);
const mkdir = promisify(fs.mkdir);

// Constants
const CONFIG = {
    CLI_ONLY: false,
    FILE_ERRORLOG: './replikate.log',
    STATE_FILE_TEMPLATE: './states/replikate_state_%SOURCE%_%TARGET%.txt',
    STATE_KEY_NAME: 'last_processed_id',
    
    LOG_OUTPUT: false,
    
    LOCAL_URL: 'http://localhost/zosciimq/index.php',
    NONCE_ROOT: './nonce/',
    QUEUE_ROOT: './queues/',
    STORE_ROOT: './store/',
    TEMP_QUEUE: 'temp/',
    FOLDER_PERMISSIONS: 0o755
};

function logError(str) {
    if (CONFIG.LOG_OUTPUT) {
        const timestamp = new Date().toISOString().replace('T', ' ').substring(0, 19);
        fs.appendFileSync(CONFIG.FILE_ERRORLOG, timestamp + ' - ' + str + '\n');
    }
}

function convertNameToBase36(strName) {
    let varResult = false;
    
    const arrParts = strName.split('-');
    const strTimestamp = arrParts[0];
    
    if (strTimestamp.length === 18) {
        const SS = strTimestamp.substring(12, 14);
        const NN = strTimestamp.substring(10, 12);
        const HH = strTimestamp.substring(8, 10);
        const DD = strTimestamp.substring(6, 8);
        const MM = strTimestamp.substring(4, 6);
        const YYYY = strTimestamp.substring(0, 4);
        const CCCC = strTimestamp.substring(14, 18);
        
        const strReorderedDecimal = SS + NN + HH + DD + MM + YYYY + CCCC;
        varResult = BigInt(strReorderedDecimal).toString(36);
    }
    
    return varResult;
}

function getRetentionFromName(strName) {
    const arrParts = strName.split('-');
    
    if (arrParts.length >= 2) {
        return parseInt(arrParts[1], 10);
    }
    
    return 0;
}

function insertSuffixBeforeExtension(strName, strSuffix) {
    const strExtension = path.extname(strName);
    const strBaseName = path.basename(strName, strExtension);
    
    if (strExtension !== null && strExtension !== '') {
        return strBaseName + strSuffix + strExtension;
    } else {
        return strBaseName + strSuffix;
    }
}

async function initFolders() {
    if (!fs.existsSync(CONFIG.QUEUE_ROOT)) {
        try {
            await mkdir(CONFIG.QUEUE_ROOT, { recursive: true, mode: CONFIG.FOLDER_PERMISSIONS });
        } catch (err) {
            logError('Fatal Could not create root queue directory: ' + CONFIG.QUEUE_ROOT);
            process.exit(1);
        }
    }
    
    if (!fs.existsSync(CONFIG.QUEUE_ROOT + CONFIG.TEMP_QUEUE)) {
        try {
            await mkdir(CONFIG.QUEUE_ROOT + CONFIG.TEMP_QUEUE, { recursive: true, mode: CONFIG.FOLDER_PERMISSIONS });
        } catch (err) {
            logError('Fatal Could not create temp queue directory: ' + CONFIG.QUEUE_ROOT + CONFIG.TEMP_QUEUE);
            process.exit(1);
        }
    }
    
    if (!fs.existsSync(CONFIG.STORE_ROOT)) {
        try {
            await mkdir(CONFIG.STORE_ROOT, { recursive: true, mode: CONFIG.FOLDER_PERMISSIONS });
        } catch (err) {
            logError('Fatal Could not create root store directory: ' + CONFIG.STORE_ROOT);
            process.exit(1);
        }
    }
    
    if (!fs.existsSync(CONFIG.NONCE_ROOT)) {
        try {
            await mkdir(CONFIG.NONCE_ROOT, { recursive: true, mode: CONFIG.FOLDER_PERMISSIONS });
        } catch (err) {
            logError('Fatal Could not create nonce directory: ' + CONFIG.NONCE_ROOT);
            process.exit(1);
        }
    }
}

function fetchNextMessage(strSourceURL, strSourceQueue, strAfterName) {
    return new Promise((resolve) => {
        const parsedUrl = url.parse(strSourceURL);
        const postData = querystring.stringify({
            action: 'fetch',
            q: strSourceQueue,
            after: strAfterName
        });
        
        const options = {
            hostname: parsedUrl.hostname,
            port: parsedUrl.port,
            path: parsedUrl.path,
            method: 'POST',
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded',
                'Content-Length': Buffer.byteLength(postData)
            },
            timeout: 10000
        };
        
        const protocol = parsedUrl.protocol === 'https:' ? https : http;
        
        const req = protocol.request(options, (res) => {
            const chunks = [];
            
            res.on('data', (chunk) => {
                chunks.push(chunk);
            });
            
            res.on('end', () => {
                const binContent = Buffer.concat(chunks);
                
                if (res.statusCode !== 200) {
                    logError(`Received HTTP Status ${res.statusCode}. Server response: ${binContent.toString()}`);
                    resolve([null, null]);
                    return;
                }
                
                // Check Content-Disposition header for filename
                const contentDisposition = res.headers['content-disposition'];
                let strName = null;
                
                if (contentDisposition) {
                    const match = contentDisposition.match(/filename="([^"]+)"/i);
                    if (match) {
                        strName = path.basename(match[1]);
                    }
                }
                
                if (strName && binContent) {
                    resolve([strName, binContent]);
                } else {
                    // Try to parse as JSON
                    try {
                        const arrJSON = JSON.parse(binContent.toString());
                        if (arrJSON.system === 'ZOSCII MQ') {
                            if (arrJSON.error && arrJSON.error.length > 0) {
                                logError(`Source MQ reported JSON error: ${arrJSON.error} - Message: ${arrJSON.message}`);
                            }
                        } else {
                            logError(`Invalid JSON. Content: ${binContent.toString().substring(0, 100)}`);
                        }
                    } catch (err) {
                        logError(`Invalid response structure (missing name or content, and not valid JSON). Content: ${binContent.toString().substring(0, 100)}`);
                    }
                    resolve([null, null]);
                }
            });
        });
        
        req.on('error', (err) => {
            logError(`cURL ERROR: Failed to connect or execute request: ${err.message}`);
            resolve([null, null]);
        });
        
        req.on('timeout', () => {
            req.destroy();
            logError('Request timeout');
            resolve([null, null]);
        });
        
        req.write(postData);
        req.end();
    });
}

function publishToQueue(strTargetQueue, intRetention, binContent) {
    return new Promise((resolve) => {
        const strRetentionDays = intRetention.toString().padStart(4, '0');
        
        const postData = querystring.stringify({
            action: 'publish',
            q: strTargetQueue,
            r: strRetentionDays,
            msg: binContent.toString('base64')
        });
        
        const parsedUrl = url.parse(CONFIG.LOCAL_URL);
        
        const options = {
            hostname: parsedUrl.hostname,
            port: parsedUrl.port || 80,
            path: parsedUrl.path,
            method: 'POST',
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded',
                'Content-Length': Buffer.byteLength(postData)
            },
            timeout: 30000
        };
        
        const protocol = parsedUrl.protocol === 'https:' ? https : http;
        
        const req = protocol.request(options, (res) => {
            const chunks = [];
            
            res.on('data', (chunk) => {
                chunks.push(chunk);
            });
            
            res.on('end', () => {
                const strResponse = Buffer.concat(chunks).toString();
                
                if (res.statusCode !== 200) {
                    logError(`Local POST to index failed with HTTP code: ${res.statusCode}`);
                    resolve(false);
                    return;
                }
                
                try {
                    const arrJSON = JSON.parse(strResponse);
                    if (arrJSON.error && arrJSON.error.length > 0) {
                        logError(`Local API Error: ${arrJSON.error} / System Error: ${arrJSON.system}`);
                        resolve(false);
                    } else {
                        resolve(true);
                    }
                } catch (err) {
                    logError(`Local POST returned invalid JSON: ${strResponse}`);
                    resolve(false);
                }
            });
        });
        
        req.on('error', (err) => {
            logError(`Exception during publish: ${err.message}`);
            resolve(false);
        });
        
        req.on('timeout', () => {
            req.destroy();
            logError('Publish timeout');
            resolve(false);
        });
        
        req.write(postData);
        req.end();
    });
}

async function saveToStore(strName, binMessage) {
    let blnResult = true;
    
    const strTempName = convertNameToBase36(strName);
    
    if (strTempName === false) {
        logError('Name conversion failed for: ' + strName);
        return false;
    }
    
    const strDir1 = strTempName.substring(0, 1);
    const strDir2 = strTempName.substring(1, 2);
    const strDir3 = strTempName.substring(2, 3);
    const strStorePath = CONFIG.STORE_ROOT + strDir1 + path.sep + strDir2 + path.sep + strDir3 + path.sep;
    
    if (!fs.existsSync(strStorePath)) {
        try {
            await mkdir(strStorePath, { recursive: true, mode: CONFIG.FOLDER_PERMISSIONS });
        } catch (err) {
            blnResult = false;
            logError('Could not create nested store directory: ' + strStorePath);
        }
    }
    
    if (blnResult) {
        strName = insertSuffixBeforeExtension(strName, '-u');
        const strFullPath = strStorePath + strName;
        
        try {
            await writeFile(strFullPath, binMessage);
        } catch (err) {
            blnResult = false;
            logError('Failed to write file to store: ' + strFullPath);
        }
    }
    
    return blnResult;
}

async function handleReplikate(strSourceURL, strSourceQueue, strTargetQueue) {
    console.log('--- Starting ZOSCII MQ Replikation ---');
    
    const strSafeSource = strSourceQueue.replace(/[^a-zA-Z0-9_-]/g, '_');
    let strSafeTarget = strTargetQueue.replace(/[^a-zA-Z0-9_-]/g, '_');
    if (strTargetQueue.length === 0) {
        strSafeTarget = 'store';
    }
    
    let strStateFilePath = CONFIG.STATE_FILE_TEMPLATE;
    strStateFilePath = strStateFilePath.replace('%SOURCE%', strSafeSource);
    strStateFilePath = strStateFilePath.replace('%TARGET%', strSafeTarget);
    
    // Ensure states directory exists
    const statesDir = path.dirname(strStateFilePath);
    if (!fs.existsSync(statesDir)) {
        await mkdir(statesDir, { recursive: true, mode: CONFIG.FOLDER_PERMISSIONS });
    }
    
    let strLastName = '';
    if (fs.existsSync(strStateFilePath)) {
        strLastName = (await readFile(strStateFilePath, 'utf8')).trim();
    }
    
    if (strLastName === '') {
        console.log('Last Processed Pointer: START');
    } else {
        console.log('Last Processed Pointer: ' + strLastName);
    }
    
    let intTotalReplikated = 0;
    
    while (true) {
        const [strName, binContent] = await fetchNextMessage(strSourceURL, strSourceQueue, strLastName);
        
        if (strName === null) {
            console.log('Source URL is caught up or returned no data. Halting this run.');
            break;
        }
        
        let blnSuccess = false;
        
        if (strTargetQueue.length > 0) {
            // Save to queue
            const intRetention = getRetentionFromName(strName);
            blnSuccess = await publishToQueue(strTargetQueue, intRetention, binContent);
            
            if (blnSuccess) {
                strLastName = strName;
                intTotalReplikated++;
                console.log('REPLIKATED: New pointer set to ' + strLastName);
            } else {
                logError('Failed to replicate message to ' + strTargetQueue + '. Halting.');
                console.log('Failed to replicate message to ' + strTargetQueue + '. Halting.');
                break;
            }
        } else {
            // Save to store
            blnSuccess = await saveToStore(strName, binContent);
            
            if (blnSuccess) {
                strLastName = strName;
                intTotalReplikated++;
            } else {
                logError('Failed to replicate message to store. Halting.');
                console.log('Failed to replicate message to store. Halting.');
                break;
            }
        }
    }
    
    await writeFile(strStateFilePath, strLastName);
    console.log('--- Replikation finished. Total messages replikated: ' + intTotalReplikated + ' ---');
}

// Entry Point
async function main() {
    // Parse command-line arguments
    let strSourceURL = '';
    let strSourceQueue = '';
    let strTargetQueue = '';
    
    process.argv.forEach(arg => {
        if (arg.startsWith('--url=')) {
            strSourceURL = arg.substring(6);
        } else if (arg.startsWith('--sq=')) {
            strSourceQueue = arg.substring(5);
        } else if (arg.startsWith('--tq=')) {
            strTargetQueue = arg.substring(5);
        }
    });
    
    if (!strSourceURL || !strSourceQueue) {
        logError('Missing required arguments.');
        console.log('Missing required arguments.');
        console.log('Usage: node replikate.js --url=<source_url> --sq=<source_queue> [--tq=<target_queue>]');
        process.exit(1);
    }
    
    await initFolders();
    await handleReplikate(strSourceURL, strSourceQueue, strTargetQueue);
}

if (require.main === module) {
    main().catch(err => {
        console.error('Error:', err);
        process.exit(1);
    });
}

module.exports = { handleReplikate, fetchNextMessage, publishToQueue, saveToStore };

File: zosciimq/csharp/replikate.cs
// Cyborg ZOSCII MQ v20251030
// (c) 2025 Cyborg Unicorn Pty Ltd.
// This software is released under MIT License.

// ZOSCII MQ Server-to-Q Replikate (Replikate.cs)
// 
// Pulls messages from a URL and publishes them to a local Target Queue.
// State is managed using a unique file per target queue, making it concurrency-safe.
// 
// Execution Example:
// Replikate.exe --url=http://other.server/index.php --sq=sourcequeue --tq=targetqueue

using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace CyborgUnicorn.ZosciiMQ
{
    public class ReplikateConstants
    {
        public const bool CLI_ONLY = false;
        public const string FILE_ERRORLOG = "./replikate.log";
        public const string STATE_FILE_TEMPLATE = "./states/replikate_state_%SOURCE%_%TARGET%.txt";
        public const string STATE_KEY_NAME = "last_processed_id";
    }
    
    public class Replikate
    {
        private static readonly HttpClient httpClient = new HttpClient();
        
        private static void LogError(string str)
        {
            if (Constants.LOG_OUTPUT)
            {
                File.AppendAllText(ReplikateConstants.FILE_ERRORLOG,
                    DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss") + " - " + str + "\n");
            }
        }
        
        private static async Task<(string name, byte[] content)> FetchNextMessage(
            string strSourceURL, string strSourceQueue, string strAfterName)
        {
            string name = null;
            byte[] content = null;
            
            try
            {
                var postData = new Dictionary<string, string>
                {
                    { "action", "fetch" },
                    { "q", strSourceQueue },
                    { "after", strAfterName }
                };
                
                var formContent = new FormUrlEncodedContent(postData);
                var response = await httpClient.PostAsync(strSourceURL, formContent);
                
                if (!response.IsSuccessStatusCode)
                {
                    LogError($"Received HTTP Status {response.StatusCode}. Server response: {await response.Content.ReadAsStringAsync()}");
                }
                else
                {
                    // Check if this is a binary response (message) or JSON (error/status)
                    var contentDisposition = response.Content.Headers.ContentDisposition;
                    
                    if (contentDisposition != null && !string.IsNullOrEmpty(contentDisposition.FileName))
                    {
                        name = contentDisposition.FileName.Trim('"');
                        name = Path.GetFileName(name);
                        content = await response.Content.ReadAsByteArrayAsync();
                    }
                    else
                    {
                        // Try to parse as JSON
                        string responseText = await response.Content.ReadAsStringAsync();
                        try
                        {
                            using (JsonDocument doc = JsonDocument.Parse(responseText))
                            {
                                JsonElement root = doc.RootElement;
                                if (root.TryGetProperty("system", out JsonElement system) && 
                                    system.GetString() == "ZOSCII MQ")
                                {
                                    if (root.TryGetProperty("error", out JsonElement error) && 
                                        !string.IsNullOrEmpty(error.GetString()))
                                    {
                                        string message = root.TryGetProperty("message", out JsonElement msg) ? 
                                            msg.GetString() : "";
                                        LogError($"Source MQ reported JSON error: {error.GetString()} - Message: {message}");
                                    }
                                }
                                else
                                {
                                    LogError($"Invalid JSON. Content: {responseText.Substring(0, Math.Min(100, responseText.Length))}");
                                }
                            }
                        }
                        catch
                        {
                            LogError($"Invalid response structure (missing name or content, and not valid JSON). Content: {responseText.Substring(0, Math.Min(100, responseText.Length))}");
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                LogError($"cURL ERROR: Failed to connect or execute request: {ex.Message}");
            }
            
            return (name, content);
        }
        
        private static async Task<bool> PublishToQueue(string strTargetQueue, int intRetention, byte[] binContent)
        {
            bool blnResult = true;
            
            // Format retention days (r) as a 4-digit string
            string strRetentionDays = intRetention.ToString("D4");
            
            try
            {
                var postData = new Dictionary<string, string>
                {
                    { "action", "publish" },
                    { "q", strTargetQueue },
                    { "r", strRetentionDays },
                    { "msg", Convert.ToBase64String(binContent) } // Base64 encode binary data
                };
                
                var formContent = new FormUrlEncodedContent(postData);
                var response = await httpClient.PostAsync(Constants.LOCAL_URL, formContent);
                
                if (!response.IsSuccessStatusCode)
                {
                    LogError($"Local POST to index failed with HTTP code: {response.StatusCode}");
                    blnResult = false;
                }
                else
                {
                    // Decode and check API response for success
                    string strResponse = await response.Content.ReadAsStringAsync();
                    
                    try
                    {
                        using (JsonDocument doc = JsonDocument.Parse(strResponse))
                        {
                            JsonElement root = doc.RootElement;
                            if (root.TryGetProperty("error", out JsonElement error))
                            {
                                string errorStr = error.GetString();
                                if (!string.IsNullOrEmpty(errorStr))
                                {
                                    string systemError = root.TryGetProperty("system", out JsonElement sys) ? 
                                        sys.GetString() : "";
                                    LogError($"Local API Error: {errorStr} / System Error: {systemError}");
                                    blnResult = false;
                                }
                            }
                        }
                    }
                    catch
                    {
                        LogError($"Local POST returned invalid JSON: {strResponse}");
                        blnResult = false;
                    }
                }
            }
            catch (Exception ex)
            {
                LogError($"Exception during publish: {ex.Message}");
                blnResult = false;
            }
            
            return blnResult;
        }
        
        private static bool SaveToStore(string strName, byte[] binMessage)
        {
            bool blnResult = true;
            
            string strTempName = Utils.ConvertNameToBase36(strName);
            
            if (strTempName == null)
            {
                LogError("Name conversion failed for: " + strName);
                return false;
            }
            
            string strDir1 = strTempName.Substring(0, 1);
            string strDir2 = strTempName.Substring(1, 1);
            string strDir3 = strTempName.Substring(2, 1);
            string strStorePath = Constants.STORE_ROOT + strDir1 + Path.DirectorySeparatorChar + 
                                strDir2 + Path.DirectorySeparatorChar + strDir3 + Path.DirectorySeparatorChar;
            
            if (!Directory.Exists(strStorePath))
            {
                try
                {
                    Directory.CreateDirectory(strStorePath);
                }
                catch
                {
                    blnResult = false;
                    LogError("Could not create nested store directory: " + strStorePath);
                }
            }
            
            if (blnResult)
            {
                strName = Utils.InsertSuffixBeforeExtension(strName, "-u");
                string strFullPath = strStorePath + strName;
                
                try
                {
                    File.WriteAllBytes(strFullPath, binMessage);
                }
                catch
                {
                    blnResult = false;
                    LogError("Failed to write file to store: " + strFullPath);
                }
            }
            
            return blnResult;
        }
        
        private static async Task HandleReplikate(string strSourceURL, string strSourceQueue, string strTargetQueue)
        {
            Console.WriteLine("--- Starting ZOSCII MQ Replikation ---");
            
            string strSafeSource = Regex.Replace(strSourceQueue, @"[^a-zA-Z0-9_-]", "_");
            string strSafeTarget = Regex.Replace(strTargetQueue, @"[^a-zA-Z0-9_-]", "_");
            if (string.IsNullOrEmpty(strTargetQueue))
            {
                strSafeTarget = "store";
            }
            
            string strStateFilePath = ReplikateConstants.STATE_FILE_TEMPLATE;
            strStateFilePath = strStateFilePath.Replace("%SOURCE%", strSafeSource);
            strStateFilePath = strStateFilePath.Replace("%TARGET%", strSafeTarget);
            
            // Ensure states directory exists
            string statesDir = Path.GetDirectoryName(strStateFilePath);
            if (!Directory.Exists(statesDir))
            {
                Directory.CreateDirectory(statesDir);
            }
            
            string strLastName = "";
            if (File.Exists(strStateFilePath))
            {
                strLastName = File.ReadAllText(strStateFilePath).Trim();
            }
            
            if (string.IsNullOrEmpty(strLastName))
            {
                Console.WriteLine("Last Processed Pointer: START");
            }
            else
            {
                Console.WriteLine("Last Processed Pointer: " + strLastName);
            }
            
            int intTotalReplikated = 0;
            
            while (true)
            {
                var (strName, binContent) = await FetchNextMessage(strSourceURL, strSourceQueue, strLastName);
                
                if (strName == null)
                {
                    Console.WriteLine("Source URL is caught up or returned no data. Halting this run.");
                    break;
                }
                
                bool blnSuccess = false;
                
                if (!string.IsNullOrEmpty(strTargetQueue))
                {
                    // Save to queue
                    int intRetention = Utils.GetRetentionFromName(strName);
                    blnSuccess = await PublishToQueue(strTargetQueue, intRetention, binContent);
                    
                    if (blnSuccess)
                    {
                        strLastName = strName;
                        intTotalReplikated++;
                        Console.WriteLine("REPLIKATED: New pointer set to " + strLastName);
                    }
                    else
                    {
                        LogError("Failed to replicate message to " + strTargetQueue + ". Halting.");
                        Console.WriteLine("Failed to replicate message to " + strTargetQueue + ". Halting.");
                        break;
                    }
                }
                else
                {
                    // Save to store
                    blnSuccess = SaveToStore(strName, binContent);
                    
                    if (blnSuccess)
                    {
                        strLastName = strName;
                        intTotalReplikated++;
                    }
                    else
                    {
                        LogError("Failed to replicate message to store. Halting.");
                        Console.WriteLine("Failed to replicate message to store. Halting.");
                        break;
                    }
                }
            }
            
            File.WriteAllText(strStateFilePath, strLastName);
            Console.WriteLine("--- Replikation finished. Total messages replikated: " + intTotalReplikated + " ---");
        }
        
        public static async Task Main(string[] args)
        {
            // Parse command-line arguments
            string strSourceURL = "";
            string strSourceQueue = "";
            string strTargetQueue = "";
            
            foreach (string arg in args)
            {
                if (arg.StartsWith("--url="))
                {
                    strSourceURL = arg.Substring(6);
                }
                else if (arg.StartsWith("--sq="))
                {
                    strSourceQueue = arg.Substring(5);
                }
                else if (arg.StartsWith("--tq="))
                {
                    strTargetQueue = arg.Substring(5);
                }
            }
            
            if (string.IsNullOrEmpty(strSourceURL) || string.IsNullOrEmpty(strSourceQueue))
            {
                LogError("Missing required arguments.");
                Console.WriteLine("Missing required arguments.");
                Console.WriteLine("Usage: Replikate --url=<source_url> --sq=<source_queue> [--tq=<target_queue>]");
                return;
            }
            
            Utils.InitFolders();
            
            await HandleReplikate(strSourceURL, strSourceQueue, strTargetQueue);
        }
    }
}